File: prototype/rails/src/app/domain/escrow/ledger_balance_test.rb
# Runs without Rails: ruby -Iapp app/domain/escrow/ledger_balance_test.rb
require "minitest/autorun"
require_relative "ledger_balance"
require_relative "ledger_movement"

module Domain
  module Escrow
    class LedgerBalanceTest < Minitest::Test
      def test_an_empty_ledger_owes_nothing
        balance = LedgerBalance.from([])

        assert_equal 0, balance.held.cents
        assert_equal 0, balance.available.cents
        assert_equal 0, balance.paid_out.cents
        refute_predicate balance, :payable?
      end

      def test_a_hold_waits_on_delivery
        balance = LedgerBalance.from([LedgerMovement.hold(Money.from_cents(40_500))])

        assert_equal 40_500, balance.held.cents
        assert_equal 0, balance.available.cents
        refute_predicate balance, :payable?
      end

      def test_a_release_moves_the_hold_to_available
        balance = LedgerBalance.from(hold_and_release(40_500))

        assert_equal 0, balance.held.cents
        assert_equal 40_500, balance.available.cents
        assert_predicate balance, :payable?
      end

      def test_a_payout_empties_what_was_available
        movements = hold_and_release(40_500) + [LedgerMovement.payout(Money.from_cents(40_500))]
        balance = LedgerBalance.from(movements)

        assert_equal 0, balance.available.cents
        assert_equal 40_500, balance.paid_out.cents
        refute_predicate balance, :payable?
      end

      def test_it_folds_a_ledger_that_holds_and_releases_more_than_once
        movements = hold_and_release(40_500) + [LedgerMovement.hold(Money.from_cents(9000))]
        balance = LedgerBalance.from(movements)

        assert_equal 9000, balance.held.cents
        assert_equal 40_500, balance.available.cents
      end

      private

      def hold_and_release(cents)
        [LedgerMovement.hold(Money.from_cents(cents)), LedgerMovement.release(Money.from_cents(cents))]
      end
    end
  end
end

File: prototype/php/src/app/Domain/Escrow/LedgerBalanceTest.php
<?php

declare(strict_types=1);

namespace App\Domain\Escrow;

use App\Domain\Money\Money;

it('holds nothing for an empty ledger', function (): void {
    $balance = LedgerBalance::from([]);

    expect($balance->held->cents)->toBe(0)
        ->and($balance->available->cents)->toBe(0)
        ->and($balance->paidOut->cents)->toBe(0)
        ->and($balance->refunded->cents)->toBe(0);
});

it('sits a hold in escrow, unpayable, with nothing refunded', function (): void {
    $balance = LedgerBalance::from([LedgerMovement::hold(Money::fromCents(9000))]);

    expect($balance->held->cents)->toBe(9000)
        ->and($balance->available->cents)->toBe(0)
        ->and($balance->paidOut->cents)->toBe(0)
        ->and($balance->refunded->cents)->toBe(0)
        ->and($balance->isPayable())->toBeFalse();
});

it('moves a release from held into available, and becomes payable', function (): void {
    $balance = LedgerBalance::from([
        LedgerMovement::hold(Money::fromCents(9000)),
        LedgerMovement::release(Money::fromCents(9000)),
    ]);

    expect($balance->held->cents)->toBe(0)
        ->and($balance->available->cents)->toBe(9000)
        ->and($balance->paidOut->cents)->toBe(0)
        ->and($balance->isPayable())->toBeTrue();
});

it('empties the available balance on payout', function (): void {
    $balance = LedgerBalance::from([
        LedgerMovement::hold(Money::fromCents(9000)),
        LedgerMovement::release(Money::fromCents(9000)),
        LedgerMovement::payout(Money::fromCents(9000)),
    ]);

    expect($balance->held->cents)->toBe(0)
        ->and($balance->available->cents)->toBe(0)
        ->and($balance->paidOut->cents)->toBe(9000);
});

it('keeps an undelivered hold apart from a paid-out release', function (): void {
    $balance = LedgerBalance::from([
        LedgerMovement::hold(Money::fromCents(9000)),
        LedgerMovement::release(Money::fromCents(9000)),
        LedgerMovement::payout(Money::fromCents(9000)),
        LedgerMovement::hold(Money::fromCents(4500)),
    ]);

    expect($balance->held->cents)->toBe(4500)
        ->and($balance->available->cents)->toBe(0)
        ->and($balance->paidOut->cents)->toBe(9000);
});

it('takes a refund before release back out of escrow, reading it as a positive amount refunded', function (): void {
    $balance = LedgerBalance::from([
        LedgerMovement::of(LedgerEntryType::Held, Money::fromCents(9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::Refunded, Money::fromCents(-9000), 'ful_1'),
    ]);

    expect($balance->held->cents)->toBe(0)
        ->and($balance->available->cents)->toBe(0)
        ->and($balance->paidOut->cents)->toBe(0)
        ->and($balance->refunded->cents)->toBe(9000);
});

it('drops the available balance on a refund after release', function (): void {
    $balance = LedgerBalance::from([
        LedgerMovement::of(LedgerEntryType::Held, Money::fromCents(9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::Released, Money::fromCents(9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::Refunded, Money::fromCents(-9000), 'ful_1'),
    ]);

    expect($balance->held->cents)->toBe(0)
        ->and($balance->available->cents)->toBe(0)
        ->and($balance->paidOut->cents)->toBe(0);
});

it('carries a negative available balance and is unpayable when a refund lands after the payout', function (): void {
    $balance = LedgerBalance::from([
        LedgerMovement::of(LedgerEntryType::Held, Money::fromCents(9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::Released, Money::fromCents(9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::PaidOut, Money::fromCents(-9000), null),
        LedgerMovement::of(LedgerEntryType::Refunded, Money::fromCents(-9000), 'ful_1'),
    ]);

    expect($balance->held->cents)->toBe(0)
        ->and($balance->available->cents)->toBe(-9000)
        ->and($balance->paidOut->cents)->toBe(9000)
        ->and($balance->isPayable())->toBeFalse();
});

it('nets a refund against its own sale rather than another still in escrow', function (): void {
    $balance = LedgerBalance::from([
        LedgerMovement::of(LedgerEntryType::Held, Money::fromCents(9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::Released, Money::fromCents(9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::Refunded, Money::fromCents(-9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::Held, Money::fromCents(5000), 'ful_2'),
    ]);

    expect($balance->held->cents)->toBe(5000)
        ->and($balance->available->cents)->toBe(0);
});

it('sums refunds across every fulfillment, whichever timing each one landed at', function (): void {
    $balance = LedgerBalance::from([
        LedgerMovement::of(LedgerEntryType::Held, Money::fromCents(9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::Refunded, Money::fromCents(-9000), 'ful_1'),
        LedgerMovement::of(LedgerEntryType::Held, Money::fromCents(4500), 'ful_2'),
        LedgerMovement::of(LedgerEntryType::Released, Money::fromCents(4500), 'ful_2'),
        LedgerMovement::of(LedgerEntryType::Refunded, Money::fromCents(-4500), 'ful_2'),
    ]);

    expect($balance->refunded->cents)->toBe(13500);
});
